base/bootstrap.go

```diff
@@ -422,25 +431,44 @@ func (cc *CouchbaseCluster) getBucket(bucketName string) (b *gocb.Bucket, teardo
 	defer cc.cachedConnectionLock.Unlock()
 
 	cacheBucket, ok := cc.cachedBucketConnections[bucketName]
+	refcountTeardown := func() {
+		cc.cachedConnectionLock.Lock()
+		defer cc.cachedConnectionLock.Unlock()
+		cc.cachedBucketConnections[bucketName].refcount--
+		if cc.cachedBucketConnections[bucketName].shouldClose {
+			cc.cachedBucketConnections[bucketName].teardownFn()
+		}
+	}
 	if ok {
-		return cacheBucket.bucket, noopTeardown, nil
+		cc.cachedBucketConnections[bucketName].refcount++
+		return cacheBucket.bucket, refcountTeardown, nil
 	}
 
 	// cached bucket not found, connect and add
 	newBucket, newTeardownFn, err := cc.connectToBucket(bucketName)
 	if err != nil {
 		return nil, nil, err
 	}
+	closeAndTeardown := func() {
+		refcount := cc.cachedBucketConnections[bucketName].refcount
+		if refcount <= 0 {
+			newTeardownFn()
+			return
+		}
+		// refcount is not zero because someone else is using bucket, so let's close
+		cc.cachedBucketConnections[bucketName].shouldClose = true
+	}
```

Comment:
Having three nested teardown functions (refcountTeardown -> closeAndTeardown -> newTeardownFn) seems a bit complex for handling `if shouldClose and refcount == 0, close the bucket and remove from cache`.  The similar naming also makes it tricky to follow exactly what's going on.  (I realize some of the bad naming existed prior to this PR)

It also makes it difficult to identify whether a particular caller is already holding cc.cachedConnectionLock.Lock().  (i.e. the refcount access on line 453 needs to be holding the lock, but I can't easily tell whether it does).

I've spent a decent chunk of time looking at this code and still am not confident there isn't a deadlock somewhere.  Any ideas for how we might simplify the amount of redirection and inline functions being used?

Reply:
I'm not at all sure I made this better, but I gave it a shot and I wrote a test for the locking behavior.